Add recordDataPtrOnStream to DeviceGuardImplInterface

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #57052 Merge CUDAFuture into ivalue::Future
* #57051 Make CUDAFuture handle any kind of device type
* #57050 Use only generic helpers in CUDAFuture
* #57049 Introduce generic MultiStreamGuard
* #57048 Simplify CUDAMultiStreamGuard
* **#57047 Add recordDataPtrOnStream to DeviceGuardImplInterface**
* #57046 Add getStreamFromPool to DeviceGuardImplInterface

We intend to merge CUDAFuture into ivalue::Future by using DeviceGuardImplInterface to avoid explicitly referring to CUDA. For that we need to add two methods to DeviceGuardImplInterface. In this PR, we add a method to record a DataPtr onto a stream with the caching allocator.

Differential Revision: [D28029161](https://our.internmc.facebook.com/intern/diff/D28029161/)

lw committed:
Add recordDataPtrOnStream to DeviceGuardImplInterface

We intend to merge CUDAFuture into ivalue::Future by using DeviceGuardImplInterface to avoid explicitly referring to CUDA. For that we need to add two methods to DeviceGuardImplInterface. In this PR, we add a method to record a DataPtr onto a stream with the caching allocator.

Differential Revision: [D28029161](https://our.internmc.facebook.com/intern/diff/D28029161/)

[ghstack-poisoned]

lw committed:
Update on "Add recordDataPtrOnStream to DeviceGuardImplInterface"


We intend to merge CUDAFuture into ivalue::Future by using DeviceGuardImplInterface to avoid explicitly referring to CUDA. For that we need to add two methods to DeviceGuardImplInterface. In this PR, we add a method to record a DataPtr onto a stream with the caching allocator.

Differential Revision: [D28029161](https://our.internmc.facebook.com/intern/diff/D28029161/)

[ghstack-poisoned]

imaginary-person commented:
`binary_linux_libtorch_3_7m_cpu_devtoolset7_shared-with-deps_build` has been failing for some other PRs as well.

lw committed:
Update on "Add recordDataPtrOnStream to DeviceGuardImplInterface"


We intend to merge CUDAFuture into ivalue::Future by using DeviceGuardImplInterface to avoid explicitly referring to CUDA. For that we need to add two methods to DeviceGuardImplInterface. In this PR, we add a method to record a DataPtr onto a stream with the caching allocator.

Differential Revision: [D28029161](https://our.internmc.facebook.com/intern/diff/D28029161/)

[ghstack-poisoned]

lw committed:
Update on "Add recordDataPtrOnStream to DeviceGuardImplInterface"


We intend to merge CUDAFuture into ivalue::Future by using DeviceGuardImplInterface to avoid explicitly referring to CUDA. For that we need to add two methods to DeviceGuardImplInterface. In this PR, we add a method to record a DataPtr onto a stream with the caching allocator.

Differential Revision: [D28029161](https://our.internmc.facebook.com/intern/diff/D28029161/)

[ghstack-poisoned]
